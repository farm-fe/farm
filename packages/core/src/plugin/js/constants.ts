export const VITE_DEFAULT_ASSETS = [
  // images
  'apng',
  'png',
  'jpe?g',
  'jfif',
  'pjpeg',
  'pjp',
  'gif',
  'svg',
  'ico',
  'webp',
  'avif',

  // media
  'mp4',
  'webm',
  'ogg',
  'mp3',
  'wav',
  'flac',
  'aac',
  'opus',
  'mov',
  'm4a',
  'vtt',

  // fonts
  'woff2?',
  'eot',
  'ttf',
  'otf',

  // other
  'webmanifest',
  'pdf',
  'txt'
];
// the name of the virtual module internal the adapter
export const VITE_ADAPTER_VIRTUAL_MODULE = 'vite-adapter-virtual:';
