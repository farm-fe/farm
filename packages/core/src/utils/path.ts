import path from 'node:path';

const windowsSlashRE = /\\/g;
export function slash(p: string): string {
  return p.replace(windowsSlashRE, '/');
}

export function withTrailingSlash(path: string): string {
  if (path[path.length - 1] !== '/') {
    return `${path}/`;
  }
  return path;
}

const postfixRE = /[?#].*$/;
export function stripQueryAndHash(path: string): string {
  return path.replace(postfixRE, '');
}

export function removeHashFromPath(path: string): string {
  const hashRE = /([_-][a-f0-9]{4,12})(\.[^./]+(\.[^./]+)*)$/;
  return path.replace(hashRE, '$2');
}

export function getShortName(file: string, root: string): string {
  const result = file.startsWith(withTrailingSlash(root))
    ? path.relative(root, file)
    : file;
  return result;
}
