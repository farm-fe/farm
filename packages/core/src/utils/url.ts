const postfixRE = /[?#].*$/;
export function cleanUrl(url: string): string {
  return url.replace(postfixRE, '');
}

const importQueryRE = /(\?|&)import=?(?:&|$)/;
export const isImportRequest = (url: string): boolean =>
  importQueryRE.test(url);

const trailingSeparatorRE = /[?&]$/;
export function removeImportQuery(url: string): string {
  return url.replace(importQueryRE, '$1').replace(trailingSeparatorRE, '');
}

export const knownJavascriptExtensionRE = /\.[tj]sx?$/;

export const urlRE = /(\?|&)url(?:&|$)/;
