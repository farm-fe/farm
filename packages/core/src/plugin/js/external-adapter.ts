import { CUSTOM_KEYS } from '../../config/constants.js';
import {
  mergeCustomExternal,
  partialExternal
} from '../../config/normalize-config/normalize-external.js';
import { UserConfig } from '../../config/types.js';
import { isArray } from '../../utils/share.js';
import { JsPlugin } from '../type.js';

/**
 * avoid add new external in config hook
 */
export function externalAdapter(): JsPlugin {
  return {
    name: 'farm:external-adapter',

    priority: -Infinity,

    config(config: UserConfig): UserConfig | Promise<UserConfig> {
      if (
        config?.compilation?.external &&
        isArray(config.compilation.external)
      ) {
        let [stringExternal, recordExternal] = mergeCustomExternal(
          config?.compilation,
          partialExternal(config.compilation.external)
        );

        return {
          compilation: {
            external: stringExternal,
            custom: {
              [CUSTOM_KEYS.external_record]: JSON.stringify(recordExternal)
            }
          }
        };
      }
    }
  };
}
