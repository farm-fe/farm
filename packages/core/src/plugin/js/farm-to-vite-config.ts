import type { UserConfig as ViteUserConfig } from 'vite';
import type { UserConfig } from '../../config/types.js';
import {
  deleteUndefinedPropertyDeeply,
  throwIncompatibleError
} from './utils.js';
import merge from '../../utils/merge.js';
import { Logger } from '../../index.js';
import { VITE_DEFAULT_ASSETS } from './constants.js';

export function farmUserConfigToViteConfig(config: UserConfig): ViteUserConfig {
  const vitePlugins = config.vitePlugins.filter(Boolean).map((plugin) => {
    if (typeof plugin === 'function') {
      return plugin().vitePlugin;
    } else {
      return plugin;
    }
  });

  let sourcemap = true;

  if (config.compilation?.sourcemap !== undefined) {
    sourcemap = Boolean(config.compilation?.sourcemap);
  }

  const viteConfig: ViteUserConfig = {
    root: config.root,
    base: config.compilation?.output?.publicPath ?? '/',
    publicDir: config.publicDir ?? 'public',
    mode: config.compilation?.mode,
    define: config.compilation?.define,
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore ignore this error
    command: config.compilation?.mode === 'production' ? 'build' : 'serve',
    resolve: {
      alias: config.compilation?.resolve?.alias,
      extensions: config.compilation?.resolve?.extensions,
      mainFields: config.compilation?.resolve?.mainFields,
      conditions: config.compilation?.resolve?.conditions,
      preserveSymlinks: config.compilation?.resolve?.symlinks === false
    },
    plugins: vitePlugins,
    server: {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore ignore error
      hmr: config.server?.hmr,
      port: config.server?.port,
      host: config.server?.host,
      strictPort: config.server?.strictPort,
      https: config.server?.https,
      proxy: config.server?.proxy as any,
      open: config.server?.open,
      watch:
        typeof config.server?.hmr === 'object'
          ? config.server.hmr?.watchOptions ?? {}
          : {}
      // other options are not supported in farm
    },
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore ignore this error
    isProduction: config.compilation?.mode === 'production',
    css: config.compilation?.css?._viteCssOptions ?? {},
    build: {
      outDir: config.compilation?.output?.path,
      sourcemap,
      minify:
        config.compilation?.minify !== undefined
          ? Boolean(config.compilation?.minify)
          : undefined,
      cssMinify:
        config.compilation?.minify !== undefined
          ? Boolean(config.compilation?.minify)
          : undefined,
      ssr: config.compilation?.output?.targetEnv === 'node',
      rollupOptions: {
        output: {
          assetFileNames: config.compilation?.output?.assetsFilename,
          entryFileNames: config.compilation?.output?.entryFilename,
          chunkFileNames: config.compilation?.output?.filename
        }
      }
      // other options are not supported in farm
    },
    // TODO make it configurable
    cacheDir: 'node_modules/.farm/cache',
    envDir: config.envDir,
    assetsInclude: [
      ...VITE_DEFAULT_ASSETS,
      ...(config.compilation?.assets?.include ?? [])
    ]
  };

  return viteConfig;
}

export function proxyViteConfig(
  viteConfig: ViteUserConfig,
  pluginName: string,
  logger: Logger
): ViteUserConfig {
  return new Proxy(viteConfig, {
    get(target, key) {
      if (typeof key !== 'string') {
        return target[key as unknown as keyof typeof target];
      }

      if (
        key === 'then' ||
        key === 'length' ||
        key === 'constructor' ||
        key === 'prototype'
      ) {
        return (target as Record<string, any>)[key];
      }

      const allowedKeys = [
        'root',
        'base',
        'publicDir',
        'mode',
        'define',
        'command',
        'resolve',
        'plugins',
        'server',
        'isProduction',
        'css',
        'build',
        'logger',
        'cacheDir',
        'envDir',
        'assetsInclude',
        // these fields are always undefined in farm
        // they are only used for compatibility
        'legacy',
        'optimizeDeps',
        'ssr',
        'logLevel',
        'experimental',
        'test',
        'clearScreen',
        'customLogger'
      ];

      if (allowedKeys.includes(String(key))) {
        if (key === 'resolve') {
          const allowedResolveKeys = [
            'alias',
            'extensions',
            'mainFields',
            'conditions',
            'preserveSymlinks',
            // farm do not set any thing for dedupe, it should always be undefined
            'dedupe'
          ];

          return new Proxy(target.resolve || {}, {
            get(resolveTarget, resolveKey) {
              if (typeof resolveKey !== 'string') {
                return target[resolveKey as unknown as keyof typeof target];
              }

              if (allowedResolveKeys.includes(String(resolveKey))) {
                return resolveTarget[resolveKey as keyof typeof resolveTarget];
              }

              if (
                resolveKey === 'then' ||
                resolveKey === 'length' ||
                resolveKey === 'constructor' ||
                resolveKey === 'prototype'
              ) {
                return (target as Record<string, any>)[key];
              }

              throw throwIncompatibleError(
                pluginName,
                'viteConfig.resolve',
                allowedResolveKeys,
                resolveKey
              );
            }
          });
        } else if (key === 'server') {
          const allowedServerKeys = [
            'hmr',
            'port',
            'host',
            'strictPort',
            'https',
            'proxy',
            'open',
            'origin',
            'watch'
          ];

          return new Proxy(target.server || {}, {
            get(serverTarget, serverKey) {
              if (typeof serverKey !== 'string') {
                return target[serverKey as unknown as keyof typeof target];
              }

              if (allowedServerKeys.includes(String(serverKey))) {
                return serverTarget[serverKey as keyof typeof serverTarget];
              }

              if (
                serverKey === 'then' ||
                serverKey === 'length' ||
                serverKey === 'constructor' ||
                serverKey === 'prototype'
              ) {
                return (target as Record<string, any>)[key];
              }

              throw throwIncompatibleError(
                pluginName,
                'viteConfig.server',
                allowedServerKeys,
                serverKey
              );
            }
          });
        } else if (key === 'css') {
          const allowedCssKeys = [
            'devSourcemap',
            'transformer',
            'modules',
            'postcss',
            'preprocessorOptions'
          ];

          return new Proxy(target.css || {}, {
            get(cssTarget, cssKey) {
              if (typeof cssKey !== 'string') {
                return target[cssKey as unknown as keyof typeof target];
              }

              if (allowedCssKeys.includes(String(cssKey))) {
                return cssTarget[cssKey as keyof typeof cssTarget];
              }

              if (
                cssKey === 'then' ||
                cssKey === 'length' ||
                cssKey === 'constructor' ||
                cssKey === 'prototype'
              ) {
                return (target as Record<string, any>)[key];
              }

              throw throwIncompatibleError(
                pluginName,
                'viteConfig.css',
                allowedCssKeys,
                cssKey
              );
            }
          });
        } else if (key === 'build') {
          const allowedBuildKeys = [
            'outDir',
            'sourcemap',
            'minify',
            'cssMinify',
            'ssr',
            'watch',
            'rollupOptions',
            'assetsDir'
          ];

          return new Proxy(target.build || {}, {
            get(buildTarget, buildKey) {
              if (typeof buildKey !== 'string') {
                return target[buildKey as unknown as keyof typeof target];
              }

              if (allowedBuildKeys.includes(String(buildKey))) {
                return buildTarget[buildKey as keyof typeof buildTarget];
              }
              if (
                buildKey === 'then' ||
                buildKey === 'length' ||
                buildKey === 'constructor' ||
                buildKey === 'prototype'
              ) {
                return (target as Record<string, any>)[key];
              }

              throw throwIncompatibleError(
                pluginName,
                'viteConfig.build',
                allowedBuildKeys,
                buildKey
              );
            }
          });
        } else if (key === 'optimizeDeps') {
          return new Proxy(target.optimizeDeps || {}, {
            get(_, optimizeDepsKey) {
              logger.warnOnce(
                `[vite-plugin] ${pluginName}: config "optimizeDeps" is not needed in farm, all of its options will be ignored. Current ignored option is: "${String(
                  optimizeDepsKey
                )}"`
              );

              if (optimizeDepsKey === 'esbuildOptions') {
                return {};
              }
              return undefined;
            }
          });
        } else if (key === 'logger') {
          return logger;
        } else if (key === 'assetsInclude') {
          return (filename: string) => {
            return (
              (viteConfig.assetsInclude as string[])?.some((r) => {
                return new RegExp(r).test(filename);
              }) ?? false
            );
          };
        }

        return target[key as keyof typeof target];
      }

      throw throwIncompatibleError(pluginName, 'viteConfig', allowedKeys, key);
    }
  });
}

export function viteConfigToFarmConfig(
  config: ViteUserConfig,
  origFarmConfig: UserConfig,
  _pluginName: string
): UserConfig {
  const farmConfig: UserConfig = {
    compilation: {}
  };

  if (config.root) {
    farmConfig.root = config.root;
  }
  if (config?.css) {
    farmConfig.compilation.css ??= {};
    farmConfig.compilation.css._viteCssOptions = config.css;
  }

  if (config.base) {
    farmConfig.compilation.output ??= {};
    farmConfig.compilation.output.publicPath = config.base;
  }
  if (config.publicDir) {
    farmConfig.publicDir = config.publicDir;
  }
  if (config.mode === 'development' || config.mode === 'production') {
    farmConfig.compilation.mode = config.mode;
  }
  if (config.define) {
    farmConfig.compilation.define = config.define;
  }
  if (config.resolve) {
    farmConfig.compilation.resolve ??= {};

    if (config.resolve.alias) {
      if (!Array.isArray(config.resolve.alias)) {
        farmConfig.compilation.resolve.alias = config.resolve.alias as Record<
          string,
          any
        >;
      } else {
        if (!farmConfig.compilation.resolve.alias) {
          farmConfig.compilation.resolve.alias = {};
        }

        const farmRegexPrefix = '$__farm_regex:';

        for (const { find, replacement } of config.resolve.alias) {
          if (find instanceof RegExp) {
            const key = farmRegexPrefix + find.source;
            farmConfig.compilation.resolve.alias[key] = replacement;
          } else {
            farmConfig.compilation.resolve.alias[find] = replacement;
          }
        }
      }
    }

    farmConfig.compilation.resolve.extensions = config.resolve.extensions;
    farmConfig.compilation.resolve.mainFields = config.resolve.mainFields;
    farmConfig.compilation.resolve.conditions = config.resolve.conditions;
    farmConfig.compilation.resolve.symlinks =
      config.resolve.preserveSymlinks != true;
  }

  if (config.server) {
    farmConfig.server ??= {};
    farmConfig.server.hmr = config.server.hmr;
    farmConfig.server.port = config.server.port;

    if (config.server.watch) {
      if (
        farmConfig.server?.hmr === true ||
        farmConfig.server?.hmr === undefined
      ) {
        farmConfig.server.hmr = {
          ...(typeof origFarmConfig?.server?.hmr === 'object'
            ? origFarmConfig.server.hmr
            : {}),
          watchOptions: config.server.watch
        };
      }
    }

    if (typeof config.server.host === 'string') {
      farmConfig.server.host = config.server.host;
    }

    farmConfig.server.strictPort = config.server.strictPort;
    farmConfig.server.https =
      typeof config.server.https === 'boolean'
        ? undefined
        : config.server.https;
    farmConfig.server.proxy = config.server.proxy as any;
    farmConfig.server.open = Boolean(config.server.open);
  }

  if (config.build) {
    farmConfig.compilation.output ??= {};
    farmConfig.compilation.output.path = config.build.outDir;

    if (
      config.build?.sourcemap !== undefined &&
      origFarmConfig.compilation?.sourcemap === undefined
    ) {
      farmConfig.compilation.sourcemap = Boolean(config.build.sourcemap);
    }

    if (config.build?.minify !== undefined) {
      farmConfig.compilation.minify = Boolean(config.build.minify);
    }

    if (
      config.build.ssr !== undefined &&
      origFarmConfig.compilation?.lazyCompilation === undefined
    ) {
      farmConfig.compilation.lazyCompilation = !config.build.ssr;
    }

    if (config.build.rollupOptions?.output !== undefined) {
      if (!Array.isArray(config.build.rollupOptions.output)) {
        const keys = ['assetFileNames', 'entryFilename', 'filename'];

        for (const k of keys) {
          /* eslint-disable @typescript-eslint/ban-ts-comment */
          // @ts-ignore type is correct
          farmConfig.compilation.output[k] =
            // @ts-ignore type is correct
            config.build.rollupOptions.output[k];
        }
      }
    }
  }

  deleteUndefinedPropertyDeeply(farmConfig);

  return merge({}, origFarmConfig, farmConfig);
}
