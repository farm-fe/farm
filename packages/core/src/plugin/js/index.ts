import merge from 'lodash.merge';
import { Config } from '../../../binding/index.js';
import {
  type JsPlugin,
  normalizeDevServerOptions,
  type UserConfig
} from '../../index.js';
import {
  DEFAULT_FILTERS,
  getCssModuleType,
  VITE_PLUGIN_DEFAULT_MODULE_TYPE
} from './utils.js';
import { VitePluginAdapter } from './vite-plugin-adapter.js';
import { existsSync, readFileSync } from 'node:fs';

// export * from './jsPluginAdapter.js';
export { VitePluginAdapter } from './vite-plugin-adapter.js';

type VitePluginType = object | (() => { vitePlugin: any; filters: string[] });
type VitePluginsType = VitePluginType[];

export function handleVitePlugins(
  vitePlugins: VitePluginsType,
  userConfig: UserConfig,
  finalConfig: Config['config']
): JsPlugin[] {
  const jsPlugins: JsPlugin[] = [];

  if (vitePlugins.length) {
    userConfig = merge({}, userConfig, {
      compilation: finalConfig,
      server: normalizeDevServerOptions(userConfig.server, finalConfig.mode)
    });
  }

  for (const vitePluginObj of vitePlugins) {
    let vitePlugin = vitePluginObj,
      filters = DEFAULT_FILTERS;

    if (typeof vitePluginObj === 'function') {
      const { vitePlugin: plugin, filters: f } = vitePluginObj();
      vitePlugin = plugin;
      filters = f;
    }
    processVitePlugin(vitePlugin, userConfig, filters, jsPlugins);
  }

  // if vitePlugins is not empty, append a load plugin to load file
  // this plugin is only for compatibility
  if (vitePlugins.length) {
    jsPlugins.push({
      name: 'farm:load',
      // has to be the last one
      priority: -100,
      load: {
        filters: {
          resolvedPaths: DEFAULT_FILTERS
        },
        executor: async (params) => {
          const { resolvedPath } = params;

          // skip lazy compiled module and non-exist file
          if (
            VitePluginAdapter.isFarmInternalVirtualModule(resolvedPath) ||
            !existsSync(resolvedPath)
          ) {
            return null;
          }

          const content = readFileSync(resolvedPath, 'utf-8');

          return {
            content,
            moduleType: VITE_PLUGIN_DEFAULT_MODULE_TYPE
          };
        }
      },
      transform: {
        filters: {
          resolvedPaths: DEFAULT_FILTERS,
          moduleTypes: []
        },
        executor: async (params) => {
          const { content, moduleId, moduleType, resolvedPath } = params;

          // skip lazy compiled module and non-exist file
          if (VitePluginAdapter.isFarmInternalVirtualModule(resolvedPath)) {
            return null;
          }
          const cssModules = finalConfig?.css?.modules?.paths ?? [
            '\\.module\\.(css|less|sass|scss)$'
          ];
          // skip css module because it will be handled by Farm
          const isCssModules = cssModules.some((reg) =>
            new RegExp(reg).test(moduleId)
          );

          // treat all scss/less/.etc lang as css
          // plugin should handle css module by itself
          if (getCssModuleType(moduleId) && !isCssModules) {
            return {
              content,
              moduleType: 'css'
            };
          }

          return {
            content,
            moduleType
          };
        }
      }
    });
  }

  return jsPlugins;
}

export function processVitePlugin(
  vitePlugin: VitePluginType,
  userConfig: UserConfig,
  filters: string[],
  jsPlugins: JsPlugin[]
) {
  const processPlugin = (plugin: any) => {
    const vitePluginAdapter = new VitePluginAdapter(
      plugin as any,
      userConfig,
      filters
    );
    convertPlugin(vitePluginAdapter);
    jsPlugins.push(vitePluginAdapter);
  };

  if (Array.isArray(vitePlugin)) {
    vitePlugin.forEach((plugin) => processPlugin(plugin));
  } else {
    processPlugin(vitePlugin);
  }
}

export function convertPlugin(plugin: JsPlugin): void {
  if (
    plugin.transform &&
    !plugin.transform.filters?.moduleTypes &&
    !plugin.transform.filters?.resolvedPaths
  ) {
    throw new Error(
      `transform hook of plugin ${plugin.name} must have at least one filter(like moduleTypes or resolvedPaths)`
    );
  }
  if (plugin.transform) {
    if (!plugin.transform.filters.moduleTypes) {
      plugin.transform.filters.moduleTypes = [];
    } else if (!plugin.transform.filters.resolvedPaths) {
      plugin.transform.filters.resolvedPaths = [];
    }
  }

  if (plugin.load?.filters?.resolvedPaths?.length) {
    if (process.platform === 'win32') {
      // replace / to \
      plugin.load.filters.resolvedPaths = plugin.load.filters.resolvedPaths.map(
        (item) => item.replaceAll('/', '\\\\')
      );
    }
  }

  if (plugin.transform?.filters?.resolvedPaths?.length) {
    if (process.platform === 'win32') {
      // replace / to \
      plugin.transform.filters.resolvedPaths =
        plugin.transform.filters.resolvedPaths.map((item) =>
          item.replaceAll('/', '\\\\')
        );
    }
  }
}
