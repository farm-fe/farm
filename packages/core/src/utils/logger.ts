import { Config } from '../../binding/index.js';
import { ColorFunction, PersistentCacheBrand, colors } from './color.js';
/* eslint-disable @typescript-eslint/no-explicit-any */
import { pad, version } from './share.js';

type LogLevelNames = 'trace' | 'debug' | 'info' | 'warn' | 'error';

enum LogLevel {
  Trace = 'trace',
  Debug = 'debug',
  Info = 'info',
  Warn = 'warn',
  Error = 'error'
}

export interface ILogger {
  trace(message: string): void;
  debug(message: string): void;
  info(message: string): void;
  warn(message: string): void;
  warnOnce(message: string): void;
  errorOnce(message: string | Error): void;
  error(message: string | Error, options?: ErrorOptions): void;
}

export interface ErrorOptions {
  exit?: boolean;
  e?: Error;
  error?: Error;
}
interface LoggerOptions {
  name?: string;
  brandColor?: ColorFunction;
  exit?: boolean;
}

const LOGGER_METHOD = {
  info: 'log',
  warn: 'warn',
  error: 'error'
} as const;

const warnOnceMessages = new Set();
const infoOnceMessages = new Set();
const errorOnceMessages = new Set();

export class Logger implements Logger {
  constructor(
    public options?: LoggerOptions,
    private levelValues: Record<LogLevelNames, number> = {
      trace: 0,
      debug: 1,
      info: 2,
      warn: 3,
      error: 4
    },
    private prefix?: string
  ) {
    if (!this.options) this.options = {};
    this.brandPrefix();
  }

  private brandPrefix(color?: (s: string | string[]) => string): void {
    const { name = 'Farm' } = this.options;
    const formattedName = colors.bold(name);
    const formattedPrefix = colors.bold(`[ ${formattedName} ]`);
    this.prefix = color ? color(formattedPrefix) : formattedPrefix;
  }

  private logMessage(
    level: LogLevelNames,
    message: string | Error,
    color?: (s: string | string[]) => string,
    showBanner = true
  ): void {
    const loggerMethod =
      level in LOGGER_METHOD
        ? LOGGER_METHOD[level as keyof typeof LOGGER_METHOD]
        : 'log';
    if (this.levelValues[level] <= this.levelValues[level]) {
      const prefix = showBanner ? this.prefix + ' ' : '';
      const loggerMessage = color ? color(prefix + message) : prefix + message;
      console[loggerMethod](loggerMessage);
    }
  }

  setPrefix(options: LoggerOptions): void {
    if (options.name) {
      this.options.name = options.name;
      this.brandPrefix(options.brandColor);
    }
  }

  trace(message: string): void {
    this.brandPrefix(colors.green);
    this.logMessage(LogLevel.Trace, message, colors.magenta);
  }

  debug(message: string): void {
    this.brandPrefix(colors.debugColor);
    this.logMessage(LogLevel.Debug, message, colors.blue);
  }

  info(message: string, iOptions?: LoggerOptions): void {
    const options: LoggerOptions | undefined = iOptions;
    if (options) {
      this.setPrefix(options);
    }
    if (!options || !options.brandColor) {
      this.brandPrefix(colors.brandColor);
    }
    this.logMessage(LogLevel.Info, message, null);
  }

  warn(message: string): void {
    this.brandPrefix(colors.yellow);
    this.logMessage(LogLevel.Warn, message, colors.yellow);
  }

  error(message: string | Error, errorOptions?: ErrorOptions): void {
    this.brandPrefix(colors.red);
    this.logMessage(LogLevel.Error, message, colors.red);

    const effectiveOptions = { ...this.options, ...errorOptions };

    if (effectiveOptions.exit) {
      process.exit(1);
    }
  }
  infoOnce(message: string) {
    if (!infoOnceMessages.has(message)) {
      infoOnceMessages.add(message);
      this.info(message);
    }
  }
  warnOnce(message: string) {
    if (!warnOnceMessages.has(message)) {
      warnOnceMessages.add(message);
      this.warn(message);
    }
  }
  errorOnce(message: string | Error) {
    if (!errorOnceMessages.has(message)) {
      errorOnceMessages.add(message);
      this.error(message);
    }
  }
  hasErrorLogged(message: string | Error) {
    return errorOnceMessages.has(message);
  }
  hasWarnLogged(message: string) {
    return warnOnceMessages.has(message);
  }
}

export function printServerUrls(
  urls: any,
  logger: Logger,
  previewFlag = false
): void {
  if (previewFlag)
    logger.info(colors.bold(colors.magenta('preview server running at: \n')));
  const colorUrl = (url: string) =>
    colors.cyan(url.replace(/:(\d+)\//, (_, port) => `:${colors.bold(port)}/`));

  const logUrl = (url: string, type: string) =>
    logger.info(
      `${colors.bold(colors.magenta('>'))} ${colors.bold(type)}${colors.bold(
        colorUrl(url)
      )}`
    );

  urls.local.map((url: string) => logUrl(url, 'Local:   '));
  urls.network.map((url: string) => logUrl(url, 'Network: '));
}

export function bootstrapLogger(options?: LoggerOptions): Logger {
  return new Logger(options);
}

export function bootstrap(times: number, config: Config) {
  const usePersistentCache = config.config.persistentCache;
  const persistentCacheFlag = usePersistentCache
    ? colors.bold(PersistentCacheBrand)
    : '';

  console.log(
    '\n',
    colors.bold(colors.brandColor(`${'ϟ'}  Farm  v${version}`))
  );
  console.log(
    `${colors.bold(colors.green(` ✓`))}  ${colors.bold(
      'Ready in'
    )} ${colors.bold(colors.green(`${times}ms`))} ${persistentCacheFlag}`,
    '\n'
  );
}

export const logger = new Logger();

export function buildErrorMessage(
  err: any,
  args: string[] = [],
  includeStack = true
): string {
  if (err.plugin) args.push(`  Plugin: ${colors.magenta(err.plugin)}`);
  const loc = err.loc ? `:${err.loc.line}:${err.loc.column}` : '';
  if (err.id) args.push(`  File: ${colors.cyan(err.id)}${loc}`);
  if (err.frame) args.push(colors.yellow(pad(err.frame)));
  if (includeStack && err.stack) args.push(pad(cleanStack(err.stack)));
  return args.join('\n');
}

function cleanStack(stack: string) {
  return stack
    .split(/\n/g)
    .filter((l) => /^\s*at/.test(l))
    .join('\n');
}
