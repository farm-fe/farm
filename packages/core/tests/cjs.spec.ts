import path from 'path';
import { fileURLToPath } from 'url';
import { expect, test } from 'vitest';

import { normalizeDevServerOptions, resolveConfig } from '../src/index.js';
import { Logger } from '../src/utils/logger.js';

test('resolveUserConfig', async () => {
  const filePath = fileURLToPath(path.dirname(import.meta.url));

  const config = await resolveConfig(
    { configPath: path.join(filePath, 'fixtures', 'config', 'farm.config.ts') },
    new Logger(),
    'development'
  );
  console.log(config.compilation.define);

  expect(config.compilation.define).toEqual({
    // FARM_HMR_HOST: true,
    // FARM_HMR_PATH: '/__hmr',
    // FARM_HMR_PORT: '9000',
    FARM_PROCESS_ENV: {
      NODE_ENV: 'development'
    },
    // FARM_HMR_PROTOCOL: 'ws',
    '$__farm_regex:(global(This)?\\.)?process\\.env\\.NODE_ENV': '"development"'
  });
  expect(config.compilation.input).toEqual({
    main: './main.tsx'
  });
  expect(config.compilation.output).toEqual({
    path: './dist',
    publicPath: '/',
    targetEnv: 'browser'
  });
  expect(config.compilation.lazyCompilation).toEqual(true);
  expect(config.compilation.sourcemap).toEqual(true);
  expect(config.compilation.minify).toEqual(false);
  expect(config.compilation.presetEnv).toEqual(false);
  expect(config.compilation.persistentCache).toEqual({
    buildDependencies: [
      // path.join(filePath, '..', 'src', 'config.ts'),
      path.join(filePath, 'fixtures', 'config', 'farm.config.ts'),
      path.join(filePath, 'fixtures', 'config', 'util.ts'),
      'module',
      'package-lock.json',
      'pnpm-lock.yaml',
      'yarn.lock'
    ],
    envs: {
      FARM_PROCESS_ENV: '{"NODE_ENV":"development"}',
      NODE_ENV: 'development',
      'package.json[name]': 'farm-fe',
      'package.json[type]': 'unknown',
      '$__farm_regex:(global(This)?\\.)?process\\.env\\.NODE_ENV':
        '"development"',
      'package.json[browser]': 'unknown',
      'package.json[exports]': 'unknown',
      'package.json[main]': 'unknown',
      'package.json[module]': 'unknown'
      // FARM_HMR_HOST: 'true',
      // FARM_HMR_PATH: '/__hmr',
      // FARM_HMR_PORT: '9000',
      // FARM_HMR_PROTOCOL: 'ws'
    },
    moduleCacheKeyStrategy: {}
  });
  expect(config.server).toEqual(
    normalizeDevServerOptions(config.server, 'development')
  );
});
