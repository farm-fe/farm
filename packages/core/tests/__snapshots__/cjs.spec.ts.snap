// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`resolveUserConfig > resolveUserConfig 1`] = `
{
  "$__farm_regex:(global(This)?\\.)?process\\.env\\.NODE_ENV": ""development"",
  "$__farm_regex:(global(This)?\\.)?process\\.env\\.mode": ""development"",
  "FARM_PROCESS_ENV": {
    "NODE_ENV": "development",
    "mode": "development",
  },
}
`;
